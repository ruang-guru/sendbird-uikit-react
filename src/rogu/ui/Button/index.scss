@import '../../../styles/variables';

.rogu-button {
  display: inline-block;
  box-shadow: none;
  border-radius: 26px;
  padding: 0px;
  font-family: var(--sendbird-font-family-default);
  cursor: pointer;
  flex: 1 1 0;

  &:hover {
    box-shadow: none;
  }

  &:focus {
    outline: none;
  }

  &:active {
    box-shadow: none;
  }

  &.rogu-button__disabled {
    cursor: not-allowed;

    @include themed() {
      background-color: t(bg-2);
    }

    &:hover {
      @include themed() {
        background-color: t(bg-2);
      }
    }
  }
}

.rogu-button--big {
  height: 40px;

  .rogu-button__text {
    margin: 10px 16px;
  }
}

.rogu-button--small {
  height: 32px;

  .rogu-button__text {
    margin: 6px 16px;
  }
}

.rogu-button--primary {
  @include themed() {
    border: 1px solid t(primary-5);
    background-color: t(primary-5);
  }

  .rogu-button__text {
    font-weight: 600;

    @include themed() {
      color: t(bg-0);
    }
  }
}

.rogu-button--secondary {
  @include themed() {
    border: 1px solid t(primary-5);
    background-color: t(bg-0);
  }


  .rogu-button__text {
    @include themed() {
      color: t(primary-5);
    }
  }

  &.rogu-button__disabled {
    cursor: not-allowed;
    border: none;

    @include themed() {
      background-color: t(bg-2);
    }

    .rogu-button__text {
      @include themed() {
        color: t(on-bg-2);
      }
    }

    &:hover {
      border: none;

      @include themed() {
        background-color: t(bg-2);
      }

      .rogu-button__text {
        @include themed() {
          color: t(on-bg-2);
        }
      }
    }
  }
}

.rogu-button--danger {
  @include themed() {
    border: 1px solid t(error--3-2);
    background-color: t(error--3-2);
  }

  .rogu-button__text {
    @include themed() {
      color: t(on-content-1);
    }
  }

  &:hover {
    @include themed() {
      background-color: t(error-4);
      border-color: t(error-4);
    }
  }

  &:focus {
    @include themed() {
      box-shadow: 0px 0px 0px 2px t(error-3);
      border-color: var(--sendbird-button-border-focus);
    }
  }

  &:active {
    box-shadow: none;

    @include themed() {
      background-color: t(error-5);
    }
  }

  &:hover &:focus {
    @include themed() {
      box-shadow: 0px 0px 0px 2px t(error-4);
    }
  }

  &:active &:focus {
    @include themed() {
      box-shadow: 0px 0px 0px 2px t(error-5);
    }
  }
}

.rogu-button--disabled {
  border: 1px solid transparent;
  cursor: not-allowed;

  @include themed() {
    background-color: t(bg-2);
    color: t(on-content-1);
  }
}
