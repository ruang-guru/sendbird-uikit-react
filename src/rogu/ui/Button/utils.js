import { Type, Size } from './type';

export function changeTypeToClassName(type) {
  switch (type) {
    case Type.PRIMARY: return 'rogu-button--primary';
    case Type.SECONDARY: return 'rogu-button--secondary';
    case Type.DANGER: return 'rogu-button--danger';
    case Type.DISABLED: return 'rogu-button--disabled';
    default: return null;
  }
}

export function changeSizeToClassName(size) {
  switch (size) {
    case Size.BIG: return 'rogu-button--big';
    case Size.SMALL: return 'rogu-button--small';
    default: return null;
  }
}
