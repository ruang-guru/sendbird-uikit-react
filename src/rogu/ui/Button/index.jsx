import React from 'react';
import PropTypes from 'prop-types';

import './index.scss';
import Label, { LabelTypography, LabelColors } from '../Label';
import { Type, Size } from './type';
import { changeTypeToClassName, changeSizeToClassName } from './utils';

export default function Button({
  className,
  type,
  size,
  children,
  disabled,
  onClick,
}) {
  const injectingClassNames = [
    ...((Array.isArray(className)) ? className : [className]),
    'rogu-button',
    (disabled ? 'rogu-button__disabled' : ''),
    changeTypeToClassName(type),
    changeSizeToClassName(size),
  ].join(' ');

  return (
    <button
      className={injectingClassNames}
      type="button"
      onClick={onClick}
      disabled={disabled}
    >
      <Label
        className="rogu-button__text"
        type={LabelTypography.BODY_3}
        color={LabelColors.ONCONTENT_1}
      >
        {children}
      </Label>
    </button>
  );
}

export const ButtonTypes = Type;
export const ButtonSizes = Size;

Button.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
  type: PropTypes.oneOf(Object.keys(Type)),
  size: PropTypes.oneOf(Object.keys(Size)),
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]),
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
};

Button.defaultProps = {
  className: '',
  type: Type.PRIMARY,
  size: Size.BIG,
  children: 'Button',
  disabled: false,
  onClick: () => { },
};
