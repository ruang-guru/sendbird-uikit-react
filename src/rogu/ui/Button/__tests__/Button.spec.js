import React from 'react';
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer';

import Button from "../index";
import { Size } from '../type';

describe('Button', () => {
  it('should do a snapshot test of the default Button DOM', function () {
    const component = renderer.create(
      <Button />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should do a snapshot test of the small Button DOM', function () {
    const component = renderer.create(
      <Button size={Size.SMALL} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should contain className', function () {
    const text = 'example-text';
    const component = shallow(<Button className={text} />);
    expect(
      component.find(".sendbird-button").hasClass(text)
    ).toBe(true);
  });
});
