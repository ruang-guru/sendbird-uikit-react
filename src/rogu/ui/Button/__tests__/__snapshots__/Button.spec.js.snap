// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should do a snapshot test of the default Button DOM 1`] = `
<button
  className=" rogu-button  rogu-button--primary rogu-button--big"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <span
    className="rogu-button__text sendbird-label rogu-label--body-3 sendbird-label--color-oncontent-1"
  >
    Button
  </span>
</button>
`;

exports[`Button should do a snapshot test of the small Button DOM 1`] = `
<button
  className=" rogu-button  rogu-button--primary rogu-button--small"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <span
    className="rogu-button__text sendbird-label rogu-label--body-3 sendbird-label--color-oncontent-1"
  >
    Button
  </span>
</button>
`;
